content/blog/en/introduction-to-testing-with-dojo.md

```diff
@@ -0,0 +1,151 @@
+
+and lead to uncaught bugs  a widgets structure means that we don't know if the action of triggering the event handler had any unexpected affects on other areas of a widget's render output.
+
+To look at how assertion templates are used, we need to update the `Profile` widget's property API to make the `username` optional.
+
+> src/widgets/Profile.tsx
+
+```tsx
+import { tsx, create } from '@dojo/framework/core/vdom';
+
+import * as css from './styles/Profile.m.css';
+
+export interface ProfileProperties {
+	username?: string;
+}
+
+const factory = create().properties<ProfileProperties>();
+
+export default factory(function Profile({ properties }) {
+	const { username = 'Stranger' } = properties();
+	return <h1 classes={[css.root]}>{`Welcome ${username}!`}</h1>;
+});
+```
+
+Firstly the `assertionTemplate` needs to be imported into our test module. Next we need create a "base" assertion that represents the widget's default render output based on the properties API, for the `Profile` it is the `h1` node with the default `username`.
+
+```tsx
+const { describe, it } = intern.getInterface('bdd');
+import harness from '@dojo/framework/testing/harness';
+import { tsx } from '@dojo/framework/core/vdom';
+import assertionTemplate from '@dojo/framework/testing/assertionTemplate';
+
+import Profile from '../../../src/widgets/Profile';
+import * as css from '../../../src/widgets/styles/Profile.m.css';
+
+const baseAssertion = assertionTemplate(() =>
+  <h1 classes={[css.root]} "assertion-key"="welcome">Welcome Stranger!</h1>
+);
+
+describe('Profile', () => {
+	it('default renders correctly', () => {
+		const h = harness(() => <Profile />);
+		h.expect(baseAssertion);
+	});
+});
+```
+
+We can test against our base assertion like we did before. In our assertion template, we add an `assertion-key` property to the node so that we can update the template for additional test scenarios.
+
+The second test scenario of the `Profile` widget is to pass the `username` property, so this means that we need a new assertion template. This could be created just like the `baseAssertion` and for this example would not be very much effort to do and maintain, however as widgets grow in complexity this would quickly become unmanageable. Instead the `assertionTemplate` [API](https://dojo.io/learn/testing/assertion-templates) can be used to factory an updated template.
```

Comment:
```suggestion
The second test scenario of the `Profile` widget is to pass the `username` property, so we need to add a new assertion template. This could be created like the `baseAssertion` and for this example would not require much effort to create and maintain. However, as widgets grow in complexity, this would quickly become unmanageable. Instead the `assertionTemplate` [API](https://dojo.io/learn/testing/assertion-templates) can be used to factory an updated template.
```